In [ ]:
import pandas as pd
data = pd.read_csv('marketing_campaign.csv', sep='\t')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

In [ ]:
data[data['Income'].isnull()]

ID  Year_Birth   Education  ... Z_CostContact  Z_Revenue  Response
10     1994        1983  Graduation  ...             3         11         0
27     5255        1986  Graduation  ...             3         11         0
43     7281        1959         PhD  ...             3         11         0
48     7244        1951  Graduation  ...             3         11         0
58     8557        1982  Graduation  ...             3         11         0
71    10629        1973    2n Cycle  ...             3         11         0
90     8996        1957         PhD  ...             3         11         0
91     9235        1957  Graduation  ...             3         11         0
92     5798        1973      Master  ...             3         11         0
128    8268        1961         PhD  ...             3         11         0
133    1295        1963  Graduation  ...             3         11         0
312    2437        1989  Graduation  ...             3         11         0
319    2863        1970  Graduation  ...             3         11         0
1379  10475        1970      Master  ...             3         11         0
1382   2902        1958  Graduation  ...             3         11         0
1383   4345        1964    2n Cycle  ...             3         11         0
1386   3769        1972         PhD  ...             3         11         0
2059   7187        1969      Master  ...             3         11         0
2061   1612        1981         PhD  ...             3         11         0
2078   5079        1971  Graduation  ...             3         11         0
2079  10339        1954      Master  ...             3         11         0
2081   3117        1955  Graduation  ...             3         11         0
2084   5250        1943      Master  ...             3         11         1
2228   8720        1978    2n Cycle  ...             3         11         0

[24 rows x 29 columns]

In [ ]:
data.sort_values(by='Recency', ascending=True)

ID  Year_Birth   Education  ... Z_CostContact  Z_Revenue  Response
66     1386        1967  Graduation  ...             3         11         0
551    5371        1989  Graduation  ...             3         11         1
347    1826        1970  Graduation  ...             3         11         1
23     4047        1954         PhD  ...             3         11         0
391    5642        1979      Master  ...             3         11         0
...     ...         ...         ...  ...           ...        ...       ...
38     8595        1973  Graduation  ...             3         11         0
444    2106        1974    2n Cycle  ...             3         11         0
1473   4070        1969         PhD  ...             3         11         1
192    7829        1900    2n Cycle  ...             3         11         0
685   10142        1976         PhD  ...             3         11         0

[2240 rows x 29 columns]

In [ ]:
data_categorical = data[['ID','Year_Birth','Education', 'Marital_Status', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response']]
data_categorical.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'AcceptedCmp3', 'AcceptedCmp4',
       'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Complain',
       'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')

In [ ]:
data_numerical = data[['Income', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',]]
data_numerical.columns

Index(['Income', 'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts',
       'MntSweetProducts', 'MntGoldProds', 'NumDealsPurchases',
       'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases',
       'NumWebVisitsMonth'],
      dtype='object')

In [ ]:
data[['Education', 'Marital_Status']].value_counts().sort_index()

Education   Marital_Status
2n Cycle    Divorced           23
            Married            81
            Single             37
            Together           57
            Widow               5
Basic       Divorced            1
            Married            20
            Single             18
            Together           14
            Widow               1
Graduation  Absurd              1
            Alone               1
            Divorced          119
            Married           433
            Single            252
            Together          286
            Widow              35
Master      Absurd              1
            Alone               1
            Divorced           37
            Married           138
            Single             75
            Together          106
            Widow              12
PhD         Alone               1
            Divorced           52
            Married           192
            Single             98
            Together 

In [ ]:
data['Marital_Status'].value_counts()

Married     864
Together    580
Single      480
Divorced    232
Widow        77
Alone         3
YOLO          2
Absurd        2
Name: Marital_Status, dtype: int64

In [ ]:
data['Year_Birth'].min()


1893

In [ ]:
data['Year_Birth'].max()

1996

In [ ]:
data.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response',
       'age'],
      dtype='object')

In [ ]:
mapper = {'ID': 'id',
          'Year_Birth': 'year_birth', 
          'Education': 'education', 
          'Marital_Status': 'marital_status', 
          'Income': 'income', 
          'Kidhome': 'kidhome',
          'Teenhome': 'teenhome', 
          'Dt_Customer': 'dt_customer', 
          'Recency': 'recency', 
          'MntWines': 'mnt_wines', 
          'MntFruits': 'mnt_fruits',
          'MntMeatProducts': 'mnt_meat_products', 
          'MntFishProducts': 'mnt_fish_products', 
          'MntSweetProducts': 'mnt_sweet_products',
          'MntGoldProds': 'mnt_gold_prods', 
          'NumDealsPurchases': 'num_deals_purchases', 
          'NumWebPurchases': 'num_web_purchases',
          'NumCatalogPurchases': 'num_catalog_purchases', 
          'NumStorePurchases': 'num_store_purchases', 
          'NumWebVisitsMonth': 'num_web_visits_month',
          'AcceptedCmp3': 'accepted_cmp_3', 
          'AcceptedCmp4': 'accepted_cmp_4', 
          'AcceptedCmp5': 'accepted_cmp_5', 
          'AcceptedCmp1': 'accepted_cmp_1',
          'AcceptedCmp2': 'accepted_cmp_2', 
          'Complain': 'complain', 
          'Z_CostContact': 'z_cost_contract', 
          'Z_Revenue': 'z_revenue', 
          'Response': 'response'}
data = data.rename(columns=mapper)
data.columns

Index(['id', 'year_birth', 'education', 'marital_status', 'income', 'kidhome',
       'teenhome', 'dt_customer', 'recency', 'mnt_wines', 'mnt_fruits',
       'mnt_meat_products', 'mnt_fish_products', 'mnt_sweet_products',
       'mnt_gold_prods', 'num_deals_purchases', 'num_web_purchases',
       'num_catalog_purchases', 'num_store_purchases', 'num_web_visits_month',
       'accepted_cmp_3', 'accepted_cmp_4', 'accepted_cmp_5', 'accepted_cmp_1',
       'accepted_cmp_2', 'complain', 'z_cost_contract', 'z_revenue',
       'response'],
      dtype='object')

In [ ]:
data[data['income'].isnull()]

id  year_birth   education  ... z_cost_contract  z_revenue  response
10     1994        1983  Graduation  ...               3         11         0
27     5255        1986  Graduation  ...               3         11         0
43     7281        1959         PhD  ...               3         11         0
48     7244        1951  Graduation  ...               3         11         0
58     8557        1982  Graduation  ...               3         11         0
71    10629        1973    2n Cycle  ...               3         11         0
90     8996        1957         PhD  ...               3         11         0
91     9235        1957  Graduation  ...               3         11         0
92     5798        1973      Master  ...               3         11         0
128    8268        1961         PhD  ...               3         11         0
133    1295        1963  Graduation  ...               3         11         0
312    2437        1989  Graduation  ...               3         11         0
319    2863        1970  Graduation  ...               3         11         0
1379  10475        1970      Master  ...               3         11         0
1382   2902        1958  Graduation  ...               3         11         0
1383   4345        1964    2n Cycle  ...               3         11         0
1386   3769        1972         PhD  ...               3         11         0
2059   7187        1969      Master  ...               3         11         0
2061   1612        1981         PhD  ...               3         11         0
2078   5079        1971  Graduation  ...               3         11         0
2079  10339        1954      Master  ...               3         11         0
2081   3117        1955  Graduation  ...               3         11         0
2084   5250        1943      Master  ...               3         11         1
2228   8720        1978    2n Cycle  ...               3         11         0

[24 rows x 29 columns]

In [ ]:
data['income'].mean()

52247.25135379061

In [ ]:
data['income'] = data['income'].fillna(data['income'].mean())
data[data['income'].isnull()]

Empty DataFrame
Columns: [id, year_birth, education, marital_status, income, kidhome, teenhome, dt_customer, recency, mnt_wines, mnt_fruits, mnt_meat_products, mnt_fish_products, mnt_sweet_products, mnt_gold_prods, num_deals_purchases, num_web_purchases, num_catalog_purchases, num_store_purchases, num_web_visits_month, accepted_cmp_3, accepted_cmp_4, accepted_cmp_5, accepted_cmp_1, accepted_cmp_2, complain, z_cost_contract, z_revenue, response]
Index: []

In [ ]:
data['income'].head(11)

0     58138.000000
1     46344.000000
2     71613.000000
3     26646.000000
4     58293.000000
5     62513.000000
6     55635.000000
7     33454.000000
8     30351.000000
9      5648.000000
10    52247.251354
Name: income, dtype: float64

In [ ]:
data['age'] = pd.DatetimeIndex(data['dt_customer']).year - data['year_birth'].astype(int)
data['age']

0       55
1       60
2       48
3       30
4       33
        ..
2235    46
2236    68
2237    33
2238    58
2239    58
Name: age, Length: 2240, dtype: int64

In [ ]:
data['age'] = pd.datetime.now().year - data['year_birth'].astype(int)
data['age']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  """Entry point for launching an IPython kernel.


0       64
1       67
2       56
3       37
4       40
        ..
2235    54
2236    75
2237    40
2238    65
2239    67
Name: age, Length: 2240, dtype: int64

In [ ]:
def age_group(age):
  if age < 21:
    age_group = 'teenager'
 
  elif age < 50:
    age_group = 'worker'
  
  else:
    age_group = 'boomer'
  return age_group
  
data['age_group'] = data['age'].apply(lambda x: age_group(x))
data['age_group'].value_counts()

worker      1476
boomer       747
teenager      17
Name: age_group, dtype: int64

In [ ]:
data[data['age_group'] == 'teenager'][['age', 'age_group']].reset_index()

index  age age_group
0      46   16  teenager
1     140   20  teenager
2     417   20  teenager
3     696   19  teenager
4     747   19  teenager
5     831   20  teenager
6     914   18  teenager
7     921   20  teenager
8     995   19  teenager
9    1163   18  teenager
10   1170   18  teenager
11   1246   20  teenager
12   1351   20  teenager
13   1850   17  teenager
14   2086   20  teenager
15   2093   19  teenager
16   2213   17  teenager

In [ ]:
data = data.sort_values(by=['recency', 'income'], ascending=[True, False]).reset_index()
data

index     id  year_birth   education  ... z_revenue  response  age  age_group
0       347   1826        1970  Graduation  ...        11         1   44     worker
1      2049   2079        1947    2n Cycle  ...        11         0   66     boomer
2      1089   7348        1958         PhD  ...        11         1   56     boomer
3      1537  10530        1959         PhD  ...        11         1   54     boomer
4      1369  10476        1958  Graduation  ...        11         0   56     boomer
...     ...    ...         ...         ...  ...       ...       ...  ...        ...
2235   1685   7947        1969  Graduation  ...        11         0   45     worker
2236    192   7829        1900    2n Cycle  ...        11         0  113     boomer
2237   1033   5263        1977    2n Cycle  ...        11         0   36     worker
2238    444   2106        1974    2n Cycle  ...        11         0   40     worker
2239   1171   3363        1974    2n Cycle  ...        11         0   40     worker

[2240 rows x 32 columns]

level_0  index     id  year_birth  ... z_revenue response  age  age_group
0           0    347   1826        1970  ...        11        1   44     worker
1           1   2049   2079        1947  ...        11        0   66     boomer
2           2   1089   7348        1958  ...        11        1   56     boomer
3           3   1537  10530        1959  ...        11        1   54     boomer
4           4   1369  10476        1958  ...        11        0   56     boomer
...       ...    ...    ...         ...  ...       ...      ...  ...        ...
2235     2235   1685   7947        1969  ...        11        0   45     worker
2236     2236    192   7829        1900  ...        11        0  113     boomer
2237     2237   1033   5263        1977  ...        11        0   36     worker
2238     2238    444   2106        1974  ...        11        0   40     worker
2239     2239   1171   3363        1974  ...        11        0   40     worker

[2240 rows x 33 columns]

Married     864
Together    580
Single      480
Divorced    232
Widow        77
Alone         3
YOLO          2
Absurd        2

In [ ]:
def marital_status_category(marital_status):
  if marital_status == 'Married':
    marital_status = 'Together'
 
  elif marital_status == 'Single' or marital_status == 'Divorced' or marital_status == 'Widow' or marital_status == 'YOLO' or marital_status == 'Absurd':
    marital_status = 'Alone'
  
  return marital_status

data['marital_status'] = data['marital_status'].apply(lambda x: marital_status_category(x))
data['marital_status'].value_counts()

Together    1444
Alone        796
Name: marital_status, dtype: int64

In [ ]:
data['mnt_fruits'].agg(['mean'])

mean    26.302232
Name: mnt_fruits, dtype: float64

In [ ]:
avg_fruit = data.groupby(by=['education', 'marital_status'])['mnt_fruits'].agg(['mean'])
avg_fruit

mean
education  marital_status           
2n Cycle   Alone           31.046154
           Together        27.971014
Basic      Alone            8.450000
           Together        12.676471
Graduation Alone           32.784314
           Together        29.634214
Master     Alone           21.880952
           Together        21.536885
PhD        Alone           20.672316
           Together        19.692557

In [ ]:
data.to_excel('customer_enhanced.xlsx', sheet_name='customer', index=False)

In [ ]:

data[data['id'] == 1994]

index    id  year_birth   education  ... z_revenue  response  age  age_group
266     10  1994        1983  Graduation  ...        11         0   30     worker

[1 rows x 32 columns]

In [ ]:
data[data['age_group'] == 'teenager']

Empty DataFrame
Columns: [id, year_birth, education, marital_status, income, kidhome, teenhome, dt_customer, recency, mnt_wines, mnt_fruits, mnt_meat_products, mnt_fish_products, mnt_sweet_products, mnt_gold_prods, num_deals_purchases, num_web_purchases, num_catalog_purchases, num_store_purchases, num_web_visits_month, accepted_cmp_3, accepted_cmp_4, accepted_cmp_5, accepted_cmp_1, accepted_cmp_2, complain, z_cost_contract, z_revenue, response, age, age_group]
Index: []